src/views/Play/Play.tsx

```diff
@@ -956,55 +968,50 @@ export class Play extends React.Component<{}, PlayState> {
 
             // the correspondence list
             <tr className="challenge-row" key="corre">
-                <span className="cell">
+                <td className="cell">
                     {_("Correspondence:")}
-                </span>
+                </td>
             </tr>,
 
             !this.anyChallengesToShow(corre_list) ?
                 <tr className="ineligible" key="corre-ineligible">
-                    <span style={{textAlign: "center"}}>
+                    <td style={{textAlign: "center"}}>
                         {this.state.show_all_challenges ?
                             _("(none)") /* translators: There are no challenges in the system, nothing to list here */ :
                             _("(none available)") /* translators: There are no challenges that this person is eligible for */}
-                    </span>
+                    </td>
                 </tr>
             :
             corre_list.map((C) => (
                 (this.visibleInChallengeList(C) || null) &&
-                    <>
+                    <React.Fragment key={C.challenge_id}>
                         {this.rengoListItem(C, user)}
-                        {(this.state.show_in_rengo_management_pane === C.challenge_id || null) &&
+                        {(this.state.show_in_rengo_management_pane.includes(C.challenge_id) || null) &&
                             this.rengoManageListItem(C, user)
                         }
-                    </>
+                    </React.Fragment>
             ))
         ];
     };
 
-    manageRengoChallenge = (C) => {
-        //console.log("managing ", C);
-        if (this.state.show_in_rengo_management_pane === C.challenge_id) {
-            this.setState({show_in_rengo_management_pane: 0});
+    toggleRengoChallengePane = (C: Challenge) => {
+        if (this.state.show_in_rengo_management_pane.includes(C.challenge_id)) {
+            this.closeChallengeManagementPane(C);
         } else {
-            this.setState({show_in_rengo_management_pane: C.challenge_id});
+            this.setState({show_in_rengo_management_pane: [C.challenge_id, ... this.state.show_in_rengo_management_pane]});
         }
     };
 
-    stopManagingRengoChallenge = () => {
-        this.setState({show_in_rengo_management_pane: 0});
-    };
-
     rengoManageListItem = (C, user) => {
         return (
-            <tr key={C.challenge_id} className={"challenge-row rengo-management-row"}>
+            <tr key={`manage-list-item-${C.challenge_id}`} className={"challenge-row rengo-management-row"}>
```

Comment:
Might be my own React ignorance, but is this key (`manage-list-item-<id>`, `rengo-list-item-<id>`) necessary? It seems like since the `React.Fragment` that wraps this is keyed on Challenge ID, there is no reason to key its children...

Reply:
Looking at the commit history, I realize this question might be for @anoek

Reply:
I think you're correct that the conversion from `<>` to `<React.Fragment>` does obviate the need for the inside keys.

Reply:
That said ... initially each of those two components were an alternate in that slot, which is why they were originally keyed by challenge id.    anoek made them both show at the same time  (very good move, why didn't I think of that!?), ay which point keying at the top became possible.

The thing is, each component is clearly designed to be instantiated in a list.   From a reuse and design perspective, doesn't this mean that they should key themselves anyhow?   I'm not sure it feels right to be putting in and taking out keys based on where something is being used...

Reply:
Yeah it's no longer necessary. Though it doesn't hurt if they are there and not needed, no need to take them out and put them back if you move stuff around or whatever.

Reply:
Perhaps this helps (from [Lists and Keys: Extracting Components with Keys](https://reactjs.org/docs/lists-and-keys.html#extracting-components-with-keys)):

>Keys only make sense in the context of the surrounding array.
>
>For example, if you extract a `ListItem` component, you should keep the key on the <ListItem /> elements in the array rather than on the <li> element in the ListItem itself.

While the code in question isn't a component (although it could be!), what I get from this is that keys should generally be directly in the `map` function if possible

Reply:
Lightglobe: you can put a key prop on anything, and React consumes it, you don't have to declare it on the component, so put the key in at the map level.   Thanks for pointing that out.   If either of us (or anyone) are in there fixing up the space thing you found, lets tidy that up too.

Reply:
.... and if those things _had_ been pulled out as proper components, instead of inserted as functions, it all would have been a lot clearer what is going on.     ` {  this.renderSometing(whatever) }` probably should be deprecated in favour of <Something whatever={whatever}>  eh?   Is that the right conclusion?

Reply:
... which is easy to say, but somethings renderSomething just wants to call many functions in the component instantiating it :s

Reply:
... I just discovered something that I can't decide whether it's hideous or wonderful:


```            corre_list.map((C) => (
                (this.visibleInChallengeList(C) || null) &&
                    <React.Fragment key={C.challenge_id}>
                        <this.rengoListItem C={C} user={user} key={`rengo-list-{C.challenge_id}`}/>
                        {(this.state.show_in_rengo_management_pane.includes(C.challenge_id) || null) &&
                            <this.rengoManageListItem C={C} user={user} key={`rengo-manage-list-{C.challenge_id}`}/>
                        }
                    </React.Fragment>
            ))
```

... local components!?   They have access to `this.someFunction()`  😮

Reply:
... so you can do this:

```
            live_list.map((C) => (
                this.visibleInChallengeList(C) ?
                    <this.rengoListItem C={C} user={user} key={`rengo-list-item-{C.challenge_id}`}/>
                : null
            )),
```

... keyed as and when needed at the map level

Reply:
Woah that's pretty funky... I like it :)

Those functions don't access any instance variables right? Maybe `static` is the way to go?  Then it would be `Play.RengoListItem` instead of `this.RengoListItem`

Reply:
Ah nvm I see you already put this in.. cool stuff!

Reply:
`Play.RengoListItem`  is nicer aesthetically - but I think it has to be `this.RengoListItem` because the reason for doing is is that you want to access functions that themselves update `state`, like `closePane` or `startChallenge` ...